eth2/operation_pool/src/lib.rs

```diff
@@ -888,15 +909,115 @@ mod tests {
+                        &bc.committee,
+                        bc.index,
+                        keypairs,
+                        i..i + step_size,
+                        slot,
+                        state,
+                        spec,
+                        if i == 0 { None } else { Some(0) },
+                    );
+                    op_pool.insert_attestation(att, state, spec).unwrap();
+                }
+            };
+
+            for committee in &committees {
+                assert_eq!(committee.committee.len(), target_committee_size);
+                // Attestations signed by only 2-3 validators
+                insert_attestations(committee, small_step_size);
+                // Attestations signed by 4+ validators
+                insert_attestations(committee, big_step_size);
+            }
+
+            let num_small = target_committee_size / small_step_size;
+            let num_big = target_committee_size / big_step_size;
+
+            assert_eq!(op_pool.attestations.read().len(), committees.len());
+            assert_eq!(
+                op_pool.num_attestations(),
+                (num_small + num_big) * committees.len()
+            );
+            assert!(op_pool.num_attestations() > max_attestations);
+
+            state.slot += spec.min_attestation_inclusion_delay;
+            let best_attestations = op_pool
+                .get_attestations(state, spec)
+                .expect("should have valid best attestations");
+            assert_eq!(best_attestations.len(), max_attestations);
+
+            let active_indices = state
+                .get_cached_active_validator_indices(RelativeEpoch::Current)
+                .unwrap();
+            let total_active_balance = state.get_total_balance(&active_indices, spec).unwrap();
+
+            // Used for asserting that rewards are in decreasing order.
+            let mut prev_reward = u64::max_value();
+
+            for att in &best_attestations {
+                // Code taken from attestation.rs, modified to sum over the attestation indices.
+                // using aggregation_bits instead of earliest_attestation_validators() for indices.
+                let fresh_validators = earliest_attestation_validators(att, state);
+                let indices = get_attesting_indices(state, &att.data, &fresh_validators).unwrap();
+
+                let rewards = indices.iter().fold(0u64, |sum, validator_index| {
+                    let reward = get_base_reward(
+                        state,
+                        *validator_index as usize,
+                        total_active_balance,
+                        spec,
+                    )
+                    .unwrap()
+                        / spec.proposer_reward_quotient;
+                    sum + reward
+                });
+
+                // Check that rewards are in decreasing order
+                assert!(prev_reward >= rewards);
+
+                prev_reward = rewards;
+            }
```

Comment:
For some reason these are not in decreasing order. Given that the implementation of `MaxCover` is pretty straightforward, I presume that something is wrong with my test case. I've been hurting my head against the wall for a couple of hours (days) but I can't seem to find it. Would you have any idea @michaelsproul ?
Once this test gets fixed, the PR will be ready for review :D

Reply:
Nice work! This test is just not taking into account that attestations included earlier in the block affect the rewards for later attestations. E.g. if we have attestations with validators `{5, 6}` then `{5, 4}`, the first one gets the reward for 5 and 6, while the second only gets the reward for 4. I added a set to keep track of the validators already rewarded, and then the test passes. Something like:

```rust
// Set of indices covered by previous attestations in `best_attestations`.
let mut seen_indices = BTreeSet::new();
// Used for asserting that rewards are in decreasing order.
let mut prev_reward = u64::max_value();

for att in &best_attestations {
    let fresh_validators_bitlist = earliest_attestation_validators(att, state);
    let att_indices =
        get_attesting_indices(state, &att.data, &fresh_validators_bitlist).unwrap();
    let fresh_indices = &att_indices - &seen_indices;

    let rewards = fresh_indices
        .iter()
        .map(|validator_index| {
            get_base_reward(
                state,
                *validator_index as usize,
                total_active_balance,
                spec,
            )
            .unwrap()
                / spec.proposer_reward_quotient
        })
        .sum();

    // Check that rewards are in decreasing order
    assert!(prev_reward >= rewards);

    prev_reward = rewards;
    seen_indices.extend(fresh_indices);
}
```

Reply:
Of course!! I knew there was something going on here, and I tried playing around with indices but didn’t quite grasp where the problem was coming from. Thanks for your help and your elegant solution! :)